tests/test_typescript.py

```diff
@@ -24,3 +29,193 @@ def test_type_alias_with_literals2():
+
+    # Case 2: unify a union with a bounded type param, which has an
+    # incompatible bound with the given union.
+    union = tst.UnionType([tst.NumberType(), tst.StringType()])
+    type_param = tp.TypeParameter("T", bound=tst.NumberType())
+
+    type_var_map = tu.unify_types(union, type_param,
+                                  tst.TypeScriptBuiltinFactory())
+    assert type_var_map == {}
+
+
+    # Case 3: unify a union with a bounded type param, which has a compatible
+    # bound with the given union.
+    type_param = tp.TypeParameter("T", bound=union)
+    type_var_map = tu.unify_types(union, type_param,
+                                  tst.TypeScriptBuiltinFactory())
+    assert type_var_map == {type_param: union}
+
+def test_union_type_unification():
+    type_param = tp.TypeParameter("T")
+    union1 = tst.UnionType([tst.NumberLiteralType(1410), tst.NumberType(), tst.StringType()])
+    union2 = tst.UnionType([type_param, tst.NumberType(), tst.StringType()])
+    assert union1.is_subtype(union2)
+
+    type_var_map = tu.unify_types(union1, union2, tst.TypeScriptBuiltinFactory())
+    assert len(type_var_map) == 1
+    assert type_var_map == {type_param: union1.types[0]}
+
+    type_param2 = tp.TypeParameter("G")
+    union3 = tst.UnionType([type_param, type_param2, tst.StringLiteralType("foo")])
+
+    type_var_map = tu.unify_types(union1, union3, tst.TypeScriptBuiltinFactory())
+    assert len(type_var_map) == 2
+    assert type_param, type_param2 in type_var_map
+    assert union1.types[1], union1.types[2] in type_var_map.values()
+
+
+def test_union_type_unification2():
+    union = tst.UnionType([tst.NumberType(), tst.StringType()])
+    assert tu.unify_types(tst.BooleanType(), union, tst.TypeScriptBuiltinFactory()) == {}
+
+    t1 = tst.NumberType()
+    t2 = tst.UnionType([tst.NumberType(), tp.TypeParameter("T")])
+    res = tu.unify_types(t1, t2, tst.TypeScriptBuiltinFactory())
+    assert len(res) == 1 and res[t2.types[1]] == t1
+
+    t1 = tst.UnionType([tst.NumberType(), tst.StringType()])
+    res = tu.unify_types(t1, t2, tst.TypeScriptBuiltinFactory())
+    assert len(res) == 1 and res[t2.types[1]] == t1.types[1]
+
+    t1 = tst.UnionType([tst.NumberType(), tst.StringLiteralType("foo"), tst.StringType()])
+    res = tu.unify_types(t1, t2, tst.TypeScriptBuiltinFactory())
+    assert len(res) == 1 and res[t2.types[1]] == t1.types[2]
+
+    t1 = tst.UnionType([tst.NumberType(), tst.NumberLiteralType(100), tst.BooleanType(), tst.StringLiteralType("foo"), tst.StringType()])
+
+    t_param1 = tp.TypeParameter("T", bound=tst.StringType())
+    helper_union = tst.UnionType([tst.BooleanType(), tst.StringType()])
+    t_param2 = tp.TypeParameter("G", bound=helper_union)
+    t2 = tst.UnionType([tst.NumberType(), t_param1, t_param2])
+
+    # Unify t1: number | 100 | boolean | "foo" | string
+    # with t2: number | T extends string | G extends (boolean | string)
+    # Result should be {T: StringType, G: BooleanType}
```

Comment:
Can you add a comment (like that one) above each test case so that it is easier for us to determine what we test?